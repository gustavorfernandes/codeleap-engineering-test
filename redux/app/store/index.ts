import { configureStore } from "@reduxjs/toolkit"
import usernameReducer from "../slices/usernameSlice"

export const store = configureStore({
  reducer: {
    username: usernameReducer
  },
});

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = ReturnType<typeof store.dispatch>
