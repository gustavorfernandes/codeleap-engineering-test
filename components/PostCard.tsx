/* eslint-disable @next/next/no-img-element */
import { IPost } from "../types/IPost"
import { formatDistanceToNow } from "date-fns"
import { useSelector } from "react-redux"
import { RootState } from "../redux/app/store"

function PostCard(post: IPost) {
  const datePost = new Date(post.created_datetime)
  const username = useSelector((state: RootState) => state.username.value)

  return (
    <div className="w-10/12 flex flex-col font-roboto border border-neutral-300 mb-4">
      <div className="w-full flex items-center justify-between bg-black py-2 px-4 text-white font-bold text-sm">
        <h2>
          {post.title}
        </h2>
        <div className={`flex items-center gap-4 justify-between ${username != post.username && "hidden"}`}>
          <button
            onClick={() => {
              console.log("Open delete modal")
            }}
          >
            <img
              className="w-[12px]"
              src="/images/delete-icon.png"
              alt="Delete"
            />
          </button>
          <button
            onClick={() => {
              console.log("Open edit modal")
            }}
          >
            <img
              className="w-[18px]"
              src="/images/edit-icon.png"
              alt="Edit"
            />
          </button>
        </div>
      </div>
      <div className="w-full flex flex-col gap-3 p-3">
        <div className="w-full flex justify-between items-center">
          <span className="font-bold text-xs text-neutral-400">
            {`@${post.username}`}
          </span>
          <span className="text-xs text-neutral-400">
            {formatDistanceToNow(datePost, { addSuffix: true })}
          </span>
        </div>
        <div className="text-xs whitespace-pre-wrap">
          {post.content}
        </div>
      </div>
    </div>
  )
}

export default PostCard
