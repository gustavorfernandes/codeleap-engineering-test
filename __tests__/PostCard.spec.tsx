import { makeStore } from '../redux/app/store/index'
import { render, screen } from "@testing-library/react"
import { Provider } from "react-redux"
import PostCard from '../components/PostCard'

describe("Post card component", () => {
  it("Edit and delete buttons must be enabled for the current user", () => {
    const store = makeStore()
    let currentUser = "John Doe"

    render(
      <Provider store={store} >
        <PostCard key={null} id={null} title={null} created_datetime={null} content={null} username={currentUser} />
      </Provider>
    )

    const editIcon = screen.getByAltText("Edit")
    const deleteIcon = screen.getByAltText("Delete")

    expect(editIcon).toBeInTheDocument()
    expect(deleteIcon).toBeInTheDocument()
  })

  it("The component must render correctly", () => {
    const store = makeStore()
    const { container } = render(
      <Provider store={store}>
        <PostCard key={null} id={null} title={null} created_datetime={null} content={null} username={null} />
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
