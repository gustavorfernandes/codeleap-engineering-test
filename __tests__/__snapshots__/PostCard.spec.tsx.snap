// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post card component The component must render correctly 1`] = `
<div>
  <div
    class="w-11/12 flex flex-col font-roboto border border-neutral-300 mb-8 sm:mb-10 md:mb-12"
  >
    <div
      class="w-full flex items-center justify-between bg-black py-2 px-4 sm:py-6 sm:px-8 sm:h-16 md:h-[4.375rem] text-white font-bold text-sm sm:text-xl md:text-[22px]"
    >
      <h2 />
      <div
        class="flex items-center gap-4 justify-between hidden"
      >
        <button>
          <img
            alt="Delete"
            class="w-[12px] sm:w-[15px] md:w-[17.5px]"
            src="/images/delete-icon.png"
          />
        </button>
        <button>
          <img
            alt="Edit"
            class="w-[18px] sm:w-[20.5px] md:w-[22.5px]"
            src="/images/edit-icon.png"
          />
        </button>
      </div>
    </div>
    <div
      class="w-full flex flex-col gap-3 p-3 sm:p-6 md:p-8"
    >
      <div
        class="w-full flex justify-between items-center"
      >
        <span
          class="font-bold text-xs sm:text-base md:text-[18px] text-neutral-400 sm:mb-1"
        >
          @null
        </span>
        <span
          class="text-xs sm:text-base md:text-[18px] text-neutral-400"
        >
          over 52 years ago
        </span>
      </div>
      <div
        class="text-xs sm:text-base md:text-[18px] whitespace-pre-wrap"
      />
    </div>
  </div>
</div>
`;
